Реализовал задания ZIO

apavlychev opened the pull request:
Использовал комбинаторы, DI, встроенные сервисы

apavlychev committed:
Реализовал 1 и 2 задания

Использовал канал ошибки и дефекты для управления исключениями, также встроенную ф-ию filterOrElse

apavlychev committed:
Реализовал 3 и 4 задания

Использовал комбинаторы <>, zipWith, zipWithPar

apavlychev committed:
Реализовал 5-ое задание

Использовал макрос @accessible (выделяет Intelijj почему-то), композиции сервисов в DI, частичное определение

avoronets84 commented on src/main/scala/module3/zio_homework/package.scala:
```diff
@@ -10,23 +20,37 @@ package object zio_homework {
    */
 
-   lazy val guessProgram = ???
+   trait Error
```
sealed trait ?

apavlychev replied:
Точно, забыл!

avoronets84 commented on src/main/scala/module3/zio_homework/package.scala:
```diff
@@ -10,23 +20,37 @@ package object zio_homework {
+
+   lazy val guessProgram: ZIO[Console with Random, Error, Unit] = for {
+     _           <- putStrLn("Угадайте число от 1 до 3:")
+     userValue   <- getStrLn.orDie.flatMap(str=>ZIO.fromOption(str.toIntOption).orElseFail(Error.ParseError("Не удалось привести к Int")))
```
здесь мне кажется немного перемудрили, for - это и так цепочка вызовов flatMap  и map в конце.
 Я бы предложил что-то типа:
```scala
(for{

        str <- zio.console.getStrLn.orDie
        int <- ZIO.fromOption(str.toIntOption)
                .mapError(_ => Error.ParseError("not parsed"))
        randomValue <- nextIntBetween(1,3)
        _ <- checkResult(int, randomValue)
      } yield ())
```
Я не рекомендую писать логику внутри for, т.к. она зачастую усложняет восприятие, поэтому предложил вынести

apavlychev replied:
Спасибо за замечания, ваш вариант гораздо приятнее. В целом иногда руки чешутся в for добавить flatMap, чтобы не было 2-ого вложенного for-а, но это явно не тот случай

Насчёт выделения логики резонно - в более сложном представлении были бы слои, и также нужно было вынести логику запроса и вывода данных через консоль выше

avoronets84 commented on src/main/scala/module3/zio_homework/package.scala:
```diff
@@ -39,30 +63,65 @@ package object zio_homework {
+      def printEffectRunningTime[R, E, A](zio: ZIO[R, E, A]):ZIO[R, E, A]
+    }
+
+    class ServiceImpl(console:Console.Service, clock:Clock.Service) extends Service {
```
я бы не рекомендовал инжектить к конструктор, как минимум консоль, т.к. будучи в сигнатуре метода она несет доп. инсайты для пользователя относительно своей работы

apavlychev replied:
Помню ваши рекомендации по поводу разделения инфраструктуры и капабилити трейтов - исправлю!

apavlychev committed:
Рефакторинг по замечаниям

Добавил sealed trait, улучшил for и убрал логику в отд. ф-ию, вынес консоль в R для сервиса
